Add recursive/limited `-a` mount & unmount | introduce `-A` for `noauto` bypass

QORTEC opened the pull request:
**Note**: 
This post was rewritten/updated.
<!--- Please fill out the following template, which will help other contributors review your Pull Request. -->

<!--- Provide a general summary of your changes in the Title above -->

<!---
Documentation on ZFS Buildbot options can be found at
https://openzfs.github.io/openzfs-docs/Developer%20Resources/Buildbot%20Options.html
-->

### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

This pull request is intended to simplify the mounting and unmounting of the desired filesystems, resolving #2901 (_recursive mounting_).

This PR modifies the `mount` and `unmount` commands by adding an optional _filesystem_ argument after the `-a` flag. This allows the user to specify a _filesystem_, and the command will mount/unmount the specified _filesystem_ and its children, provided they are 'available'.
The code for _recursive_/limited mounting uses the `-a` code/logic paths. In practice, this means that the user can specify any valid _filesystem_, even those with the `canmount=off` property. The code will then run the standard `-a` code with the additional limitation that the dataset must be the specified _filesystem_ or its children.

This PR also introduces the `-A` flag with the goal of simplifying the mounting and unmounting of datasets with the `canmount=noauto` property. The `-A` flag is a modified version of `-a`, and the code uses the same code paths that `-a` uses. However, in this case, we bypass the `canmount=noauto` check, resulting in file filesystems being treated as `canmount=yes`. To maintain the expected behavior, that datasets with `canmount=noauto` should not be automatically mounted unless explicitly specified, the `-A` flag requires a _filesystem_ to be specified.

### Description
<!--- Describe your changes in detail -->

I implemented _recursive_/limited mounting by leveraging the `zfs mount -a` code that was already present:
- `share_mount()` code is called by `zfs mount`
  - `do_all` was originally an `int` but was used as a true or false switch, so I changed it into a more appropriate `boolean_t`
  - I introduced `boolean_t do_noauto`, which is linked to the `-A` flag (for mounting datasets with `noauto`)
  - I introduced `const char *filesystem`, which takes input from `argv[0]`
  - We validate that `filesystem` is a valid ZFS filesystem by using `zfs_open()`
  - Then we pass the `filesystem` to `get_all_datasets()` using the `parent` parameter
- `get_all_datasets()` returns a list of datasets (to mount)
  - The `const char *parent` parameter was added, and it's the value passed to `ga_parent` in the `get_all_state_t` struct
  - `zfs_iter_root()` iterates over root datasets, and pass them to `get_one_dataset()`
- `get_all_state_t` struct is used to pass arguments to `get_one_dataset()`
  - `const char *ga_parent` was added; it's used for the parent dataset to match against
- `get_one_dataset()` iterates over and returns valid datasets (to mount)
  - We added an extra check: if `ga_parent` has a value, we pass the value to `zfs_related_dataset()` to check if the datasets are related
    - If the dataset is related, then continue
    - If the dataset is not related, then proceed to the next iteration
- `share_mount()` At this point, we have a list of datasets that start with or is `filesystem` itself
  - Next, we pass the value of `do_noauto` to `sm_mount_noauto` in the `share_mount_state_t` struct
  - `share_mount_one_cb()` is called to mount the list of datasets
- `share_mount_state_t` struct is used to pass arguments to `share_mount_one_cb()`
  - `boolean_t sm_mount_noauto` was added; it's used to treat `canmount=noauto` as `canmount=on`
- `share_mount_one_cb()` is a callback for mounting filesystems
  - It calls `share_mount_one()` and passes the value `sm_mount_noauto` to the `mount_noauto` property
- `share_mount_one()` is responsible for mounting filesystems
  - `boolean_t mount_noauto` was added so we can override the `canmount` check to treat `canmount=noauto` as `canmount=on`
  - The `canmount` check was modified to include `mount_noauto`
    - If `mount_noauto` is false, use normal logic
    - If `mount_noauto` is true, treat it as an _explicit_ mount/`canmount=yes`


I implemented _recursive_/limited unmounting by leveraging the `zfs unmount -a` code that was already present:
- `unshare_unmount()` code is called by `zfs unmount`
  - `do_all` was originally an `int` but was used as a true or false switch, so I changed it into a more appropriate `boolean_t`
  - I introduced `boolean_t do_noauto`, which is linked to the `-A` flag (for mounting datasets with `noauto`)
  - I introduced `const char *filesystem`, which takes input from `argv[0]`
  - We validate that `filesystem` is a valid ZFS filesystem by using `zfs_open()`
  - The `canmount` check was modified to include `mount_noauto`
    - If `mount_noauto` is false, use normal logic
    - If `mount_noauto` is true, treat it as an mount/`canmount=yes`
  - We added an extra check: if `filesystem` has a value, we pass the value to `zfs_related_dataset()` to check if the datasets are related
    - If the dataset is related, then continue
    - If the dataset is not related, then proceed to the next iteration

`libzfs.h`

---

Added the following functions:
- `related_dataset()`: Checks if the datasets are the same or if the second one is a child
  - Check if both datasets are the same
    - If true: `(return B_TRUE)`
  - Check if it is a child dataset using `is_descendant()`
    - If true: `(return B_TRUE)`
  - `(return B_FALSE)`
- `zfs_related_dataset()`: A helper function for `related_dataset()`

`libzfs_dataset.c`

---

zfs-mount(8):

**zfs mount**
&nbsp;-a _filesystem_
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Mount the specified filesystem and its children, provided they are available.
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;If no filesystem is specified, then all available ZFS file systems are mounted; may be invoked automatically as part of the boot process if configured.
&nbsp;-A _filesystem_
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Mount the specified filesystem and its children, provided they are available.
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Note: datasets with the `canmount=noauto` property will also be mounted.

**zfs unmount**
&nbsp;-a _filesystem_
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Unmount the specified filesystem and its children, provided they are available.
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;If no filesystem is specified, then all available ZFS file systems are unmounted. Invoked automatically as part of the shutdown process.
&nbsp;-A _filesystem_
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Unmount the specified filesystem and its children, provided they are available.
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Note: datasets with the `canmount=noauto` property will also be unmounted.

---

**Note**:
This pull-request should be carefully reviewed before it's accepted:
- I AM NOT a `c` programmer; this is the first `c` project that I've worked on.
- I am using ChatGPT to:
  - Get/verify my understanding of what the code is doing.
  - Request specific code examples, such as:
    - "compare two strings"
    - "zfs function that retrieves the dataset name"
  - Review code, looking for errors.
- I do interrogate ChatGPT about any results that I don't understand.

I _feel_ like I have a basic/limited understanding of the ZFS code I've touched and have made an effort to keep my changes as simple and low-impact as possible.
The implemented code is simple and easy to understand. However, it is important to note that since I am not a `C` developer, and that my code should be carefully reviewed for common beginner-level mistakes, such as memory management.

In its current state, this PR should be ready for review. 

### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
<!--- If your change is a performance enhancement, please provide benchmarks here. -->
<!--- Please think about using the draft PR feature if appropriate -->
I mounted and unmounted datasets on a test PC running Linux.

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [x] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [x] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [ ] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

ahesford commented:
Wow. If somebody with no programming knowledge submitted an AI-generated PR on one of my repos, I'd ban the person for life.

Offer fixes and features you've thoughtfully developed, not that you shoved into a chatbot.

endgame commented:
As much as LLMs worry me, @QORTEC seems to have used them to gain an understanding of the code and asked for specific low-level coding advice like "compare two strings". This seems like the class of question which would be perfectly reasonable to search for on StackOverflow or other help sites.

I think it would be a different story if @QORTEC had generated the patch wholesale. Also, extreme negative reactions are just going to make people hide their LLM usage; I think @QORTEC definitely did the right thing here by explicitly calling it out as a good-faith attempt and his best effort, so that the PR could be reviewed thoroughly.

endgame commented on cmd/zfs/zfs_main.c:
```diff
@@ -307,7 +307,7 @@ get_usage(zfs_help_t idx)
 	case HELP_MOUNT:
 		return (gettext("\tmount\n"
-		    "\tmount [-flvO] [-o opts] <-a | filesystem>\n"));
+		    "\tmount [-flvO] [-o opts] -r <-a | filesystem>\n"));
```
Maybe include `-r` with other short ops?
```suggestion
		    "\tmount [-flrvO] [-o opts] <-a | filesystem>\n"));
```
Or, if it's special enough, bracket it to make clear that it's optional?
```suggestion
		    "\tmount [-flvO] [-o opts] [-r] <-a | filesystem>\n"));
```

QORTEC replied:
Yep, this line really needs fixing:

`-r` requires `filesystem`  so it should be `[-r filesystem]`
`-a` has an optional `filesystem`, does this look correct?  `<-a [filesystem] | filesystem>`

endgame commented on cmd/zfs/zfs_main.c:
```diff
@@ -6769,18 +6770,29 @@ get_one_dataset(zfs_handle_t *zhp, void *data)
+	/*
+	 * Skip any dataset whos name doesn't have a prefix matching ga_dataset.
+	 */
+	if (state->ga_dataset != NULL && strncmp(zfs_get_name(zhp), 
```
Not sure what `zfs_get_name(zhp)` returns, but do you need an additional check that `strlen(state->ga_dataset) < strlen(zfs_get_name(zhp))`?

lundman replied:
I am guessing, but the use of `strlen()` in there is probably looking for datasets beneath a mountpoint. So if it starts with `ga_dataset` go and process it. `strncmp("/mount/point/lower", "/mount/point", 12)`

QORTEC replied:
`zfs_get_name(zhp)`: I understand it to return the datasets's path `zpool/path/to/dataset` (AI)
`strncmp()`: _compares two strings character by character_
`strlen()` is used limit the compare to the length of `state->ga_dataset` (basically a prefix match)

The logic being that if you want to mount everything under `zpool/path` recursively
 - check the name of the current dataset
 - if the datasets name starts with `zpool/path` then continue (with the function) 
 - else `return (0)` (discard this iteration)

endgame commented on cmd/zfs/zfs_main.c:
```diff
@@ -6808,6 +6820,7 @@ typedef struct share_mount_state {
 	uint_t	sm_total; /* number of filesystems to process */
 	uint_t	sm_done; /* number of filesystems processed */
 	int	sm_status; /* -1 if any of the share/mount operations failed */
+	boolean_t	sm_explicit;
```
Does this struct need re-aligning, and does this field need a comment? Also, should this be called `sm_recursive` instead of `sm_explicit`?

QORTEC replied:
Thanks!

endgame commented on cmd/zfs/zfs_main.c:
```diff
@@ -6769,18 +6770,29 @@ get_one_dataset(zfs_handle_t *zhp, void *data)
 	}
+
+	/*
+	 * Skip any dataset whos name doesn't have a prefix matching ga_dataset.
```
```suggestion
	 * Skip any dataset whose name doesn't have a prefix matching ga_dataset.
```

endgame commented on cmd/zfs/zfs_main.c:
```diff
@@ -6731,6 +6731,7 @@ zfs_do_holds(int argc, char **argv)
 typedef struct get_all_state {
 	boolean_t	ga_verbose;
 	get_all_cb_t	*ga_cbp;
+	const char	*ga_dataset;
```
Who owns this pointer, and do we have to worry about it being deallocated from underneath us?

ahesford replied:
This pointer is set to the value of a command-line argument, so the `exec` system call that launched the process owns it. It will not be deallocated for the life of the program.

QORTEC replied:
Thanks @ahesford

endgame reviewed (commented):
I know C but I don't work on OpenZFS. HTH.

ahesford commented:
I certainly appreciate that the submitter was honest about relying on ChatGPT, but the statements "I do not know C", "I used ChatGPT to get a[n] understanding", "I requested specific code", and "I had ChatGPT review the finalized code" is functionally equivalent to deferring all judgment about the changeset to a large language model.

Even in a piece of software less important than a filesystem people rely on for robust data storage, deferring judgment about patches to a system devoid of sentience or an understanding of the domain, with a known history of complete fabrication, is undesirable. Contributions of this nature can sap project resources as more effort must be expended on review and correction. Blind submissions of this nature should be discouraged; those without a thorough understanding of the code they are touching, or even of the language in which it's written, should instead be encouraged to contribute in more meaningful ways. Even a feature-request issue with a well-formed description of desired behavior is more helpful than copying and pasting code snippets.

It would be almost as bad to submit PRs consisting of functions naively copied from StackOverflow.

ahrens commented:
@ahesford 
> Contributions of this nature can sap project resources as more effort must be expended on review and correction.

You are free to engage with the OpenZFS project as much or as little as you like, including not expending your resources on reviewing this PR.  Please allow others to contribute as they would like, including reviewing this PR and helping the submitter to improve it.  Please do not attempt to speak on the behalf of the project to discourage others contributions.

ahesford commented:
I never attempted to speak on behalf of this project. To the contrary, I highlighted the independence of my opinion by noting how I would react to similar submissions on my own repositories. It is also abundantly clear from the badges on my comments that I am an observer rather than a member of the organization.

My initial and subsequent comments were intended as a (harshly) negative review, suggesting that this class of code contributions has substantial drawback. If you believe these comments are inappropriate, you of course have editorial authority.

ahrens commented:
@QORTEC Thank you for working to improve OpenZFS!  I hope you'll take a look at the code review comments made by @endgame and @lundman and continue refining this PR.

I've created the separate discussion topic https://github.com/openzfs/zfs/discussions/15269 to share my thoughts on AI-assisted contributions.

kgraaf commented:
While this is a relatively simple and minor set of changes, the very nature of open source development clearly relies on the baseline assumption that anyone submitting PRs generally knows what they're doing. The creator of this one, by their very own admission, does not. It's pretty difficult for me to see submissions like this one as being made in a good faith effort: there is no shortage of materials on learning C (of all languages) which the creator appears to have made no effort whatsoever to engage with. "Judging PRs by their merits" sounds nice, but are maintainers of open source projects now supposed to assume that every contribution potentially contain machine-generated sludge with no thought put into it? I think that says little good about the spirit of collaboration.
 
Programming languages are, at their essence, [tools of thought](https://www.eecg.utoronto.ca/~jzhu/csc326/readings/iverson.pdf), and thinking cannot be delegated to text prediction models.

endgame commented:
> open source development clearly relies on the baseline assumption that anyone submitting PRs generally knows what they're doing on some level. The creator of this one, by their very own admission, does not.

I started from the assumption that OP was **a** programmer, but not a **C** programmer, and from there assumed familiarity with the execution model of imperative languages but not necessarily the C libraries. I'm not sure why other commenters did not. I think the careful formatting of the opening statement led me to form this view, though that can of course be generated by an LLM in seconds.

ahesford commented:
Sigh. Let me start with a substantive criticism of the change submitted and add a few more general comments after.

1. The fundamental logic in this PR is fatally flawed. The parent dataset taken as input is completely unsanitized (not to mention foolishly assigned: `const char *dataset = (argv[0] != NULL) ? argv[0] : NULL;` is a *really* goofy way to just say `const char *dataset = arvg[0]`). Later, that unsanitized prefix is used as a prefix against which to match the list of all datasets. The prefix doesn't even have to be a complete dataset name, and the check is completely divorced from the actual hierarchy! Specifying a dataset argument of `pool/a` will match `pool/a/b` as well as `pool/another`, contrary to the spirit of the addition. It even looks like specifying an argument of `p` will match `pool1` as well as `pool2`.
    [Note: this assumes that `zfs_get_name` even returns a fully qualified path and not just the name **of that node**, *e.g.*, returning `child` for `pool/parent/child`. I haven't looked closely enough to be sure.]
2. The added inline help is fundamentally wrong and also breaks convention with respect to the difference between `[optional]` and `<required>` parameters.

The first issue is a critical flaw that should have disqualified this submission out of the gate. Instead, an uniformed submission has caused people to devote valuable time to understanding a changeset *that doesn't even do what it says*.

Whether or not the submitter is a programmer, it was declared up front that "I *feel* like I basic/limited have a understanding on the code I touched", which runs counter to the expression of a "baseline assumption that anyone submitting PRs generally knows what they're doing on some level" by @kgraaf as well as the sentiment expressed by @ahrens in #15269: "More importantly, we would like for submitters to have the time and understanding to participate in the code review process, and ideally to be around to help with problems that may arise with their code post-integration." Both of these statements are reasonable and I agree with each of them. 

While there is nothing wrong with consulting "AI" *as a reference* by somebody with a modicum of understanding about the patch being submitted, that is very different than what has been done here.

endgame commented:
Thanks for the thorough explanation.

QORTEC commented:
> @QORTEC seems to have used them [LLMs] to gain an understanding of the code and asked for specific low-level coding advice like "compare two strings".

> I think it would be a different story if @QORTEC had generated the patch wholesale

> thinking cannot be delegated to text prediction models

The issue with LLMs is that they don't _think_, so its is on us (users of LLMs) to do the thinking for them, and then decide if the output is valid or not;
- If you give an LLM a broad question like to "implement recursive mounting in ZFS", it will give you tons of useless _sludge_. 
- If you point it at the correct source code saying "implement recursive mounting in ZFS", it will attempt to do so, but even if is successful it probably break you source.
- If you ask it to "compare two strings" now that's something LLMs can easily answer much more quickly and accurately.
- A separate issue being that LLMs wont always choose the correct (being the most simplest or efficient) method.

> statements [... are] functionally equivalent to deferring all judgment about the changeset to a large language model

I think this is fair reaction to those statements.

> there is no shortage of materials on learning C (of all languages) which the creator appears to have made no effort whatsoever to engage with

> [open source development has] baseline assumption that anyone submitting PRs generally knows what they're doing on some level

Fair, though it's not so much as I never took a look at `c` because I did. It was difficult to get into, and in the end I decided use other languages with a lower bar to entry.
I find it difficult to learn in a structured programming class when I have no experience in the language, but it's also difficult to learn on my own (with only reference material) when there any where/one to ask questions. My approach to learning new languages is to work towards a task, and learning on they way what is needed to accomplish said task.
I personally don't like to claim knowledge when I don't _feel_ like I can defend said knowledge, and everything I know when it comes to said languages tends to be narrow in scope.

> I started from the assumption that OP was a programmer, but not a C programmer

> I think the careful formatting of the opening statement led me to form this view, though that can of course be generated by an LLM in seconds.

If I must claim knowledge;
- I have taken a `Visual Basic` class, in which I didn't really learn anything.
- I also had another class with a scripting language, but I don't even remember what language it was.
- BASH, use it for most tasks I don't want to do my self...
- python, it's an easy language to get into, and that I'm most proficient in (after BASH)

Of course I've used/played with others, but probably only enought to claim knowledge on then their syntax and some basic commands.

> Even a feature-request issue with a well-formed description of desired behavior is more helpful than copying and pasting code snippets.

Issue #2901, is almost a decade old and you'll notice it has been buried and no one is really interested in working on it.

> suggesting that this class of code contributions has substantial drawback

> are maintainers of open source projects now supposed to assume that every contribution potentially contain machine-generated sludge with no thought put into it?

> extreme negative reactions are just going to make people hide their LLM usage

> I've created the separate discussion topic #15269 to share my thoughts on AI-assisted contributions.

I think this is an important issue to discuss, and perhaps it's over due.

> My initial and subsequent comments were intended as a (harshly) negative review

Honestly, besides you initial comment which feels like it was probably a knee jerk reaction, I think that your reviews while _harsh_ are fair, and you explained your options/thoughts well.

> @QORTEC Thank you for working to improve OpenZFS! I hope you'll take a look at the code review comments made by [@]endgame and [@]lundman and continue refining this PR.

Thank's for the kind words

---

In closing the purpose of this PR was to bring attention to a decade old issue #2901, and to say "Hay, at least I tried. Is my implementation any good?"
The reason I specifically called out my lack of confidence the way I did, was because I know you can do some really stupid stuff in `c` and I don't know if I'm guilty if it.
Personally, at best I would equate my level of `c` understanding to be at a _Introduction to C_ level.

Also the fact that AI is becoming easier to access (like most new technology), makes it possible for people with "less" knowledge to work on a level that is beyond their personal skill set. 
How ZFS chooses to treat AI, and AI assisted submissions is an important discussion that needs to happen.

No, I didn't use any AI for this response **only** spell check

Quotes from:
@ahesford @endgame @kgraaf @ahrens

QORTEC commented on cmd/zfs/zfs_main.c:
```diff
@@ -6769,18 +6770,29 @@ get_one_dataset(zfs_handle_t *zhp, void *data)
 	}
+
+	/*
+	 * Skip any dataset whos name doesn't have a prefix matching ga_dataset.
```
Thanks

ahesford commented:
I'll defer to people with an understanding of ZFS internals to suggest the best method for checking lineage of a filesystem, but *if* string comparisons on the output of `zfs_get_name` are sufficient, the following changes should be made to make sure that the added features perform as expected:
- All references to `dataset` should be changed to `filesystem`; share operations only act on filesystems (at least as documented), and recursive mount operations only make sense for filesystems. (While you can certainly mount a snapshot, a snapshot can't be regarded as the parent of anything mountable.) This is really just a matter of clarity.
- When the `dataset` (now `filesystem`) parameter is taken from the command line, it should be validated by ensuring that it does not end in a trailing slash and that it actually names an existing filesystem. If these conditions fail, a meaningful error message should be reported rather than further attempting to walk the tree.
- A failed `strncmp` comparison is sufficient to reject a filesystem during the mount/share walk, but a successful comparison is not sufficient to accept it. The filesystem being tested against the parent should have a name that is either wholly identical to the parent or whose first character *after the matching prefix* is a slash.
- The help text and manual should be fixed to properly describe optional or required parameters. Something like
    ````
    usage:
    	mount
    	mount [-flvO] [-o options] -a [filesystem]
    	mount [-flvrO] [-o options] filesystem
    ```
    would be a bit more clear that `r` and `a` are mutually exclusive and that the `filesystem` parameter is only optional when `-a` is specified, but there may be a more concise way to describe these concepts.
- These recursive options also apply to `zfs share`, so the usage and manual page for that command should be similarly modified.

QORTEC commented on cmd/zfs/zfs_main.c:
```diff
@@ -7195,14 +7212,23 @@ share_mount(int op, int argc, char **argv)
+			usage(B_FALSE);
+		}
+
+		const char *dataset = (argv[0] != NULL) ? argv[0] : NULL;
```
@ahesford raised 3 points regarding dataset

that the line should read const char *dataset = arvg[0]
we need to add a check to ensure dataset, is in fact actually a valid (existing) dataset
to improve clarity all instances dataset should be changed to filesystem to be inline with documentation

ahesford commented on cmd/zfs/zfs_main.c:
```diff
@@ -6772,12 +6773,31 @@ get_one_dataset(zfs_handle_t *zhp, void *data)
+		/*
+		 * Skip any dataset with the canmount=off property.
+		 */
+		if (zfs_prop_get_int(zhp, ZFS_PROP_CANMOUNT) == ZFS_CANMOUNT_OFF) {
+			zfs_close(zhp);
+			return (0);
+		}
+
```
Why is this here? There are checks elsewhere to exclude filesystems with `canmount=off`, so there's no need to duplicate the logic here.

QORTEC replied:
Without this check `-r` will explicitly try mounting `canmount=off` which results in an error.
The check you mentioned is in the mount code.

ahesford replied:
The right place for that check is not here, because it's changing the notional behavior of `get_one_filesystem` beyond your intended scope of "descending from this point". Having the function enumerate filesystems that *include* `canmount=off` when no parent is specified but *omit* the same just because a desired parent was specified is counter-intuitive.

The `canmount` check *still* belongs in `share_mount_one`, but it highlights an inconsistency in your definitions of scope-limited `-a` and `-r`. The idea behind `mount -a` is "mount everything that can be mounted", while an explicit mount is supposed to ignore things like `canmount=noauto` because it's been specifically requested. Nevertheless, `canmount=off` is a hard stop that triggers an error.

Presumably, you introduced `-r` to do the equivalent of

```sh
for fs in <children of some ZFS filesystem>; do
    zfs mount $fs
done
```

but what is the right behavior if you do this and one of the filesystems expressly disallows mounting with `canmount=off`? Is it to ignore the error and continue, or throw an error? The former means that `-r` is a sort of weird `-a` that behaves exactly as `-a` except that it ignores `canmount=noauto` and mounts it anyway. And yet using recursion is effectively an "automatic" operation. However, failing is also problematic, because then it leaves some subset of the desired mounts in a mounted state.

I'd submit that the whole notion of `-r` is ill-formed, and you should abandon the new argument and stick to just adding the option for limiting the scope of `-a` to some subtree. If not, you'll need to separate the logic between `explicit` and `recursive` so you can properly ignore `canmount=off` for recursive operations as well.

ahesford commented on cmd/zfs/zfs_main.c:
```diff
@@ -6772,12 +6773,31 @@ get_one_dataset(zfs_handle_t *zhp, void *data)
+		 * Skip any dataset whose name doesn't match ga_filesystem exactly
+		 * or have a prefix matching filesystem.
+		 */
+		if (!strcmp(zfs_get_name(zhp), state->ga_filesystem) == 0 ||
+		    !strncmp(zfs_get_name(zhp), filesystem, strlen(filesystem) == 0)) {
+			zfs_close(zhp);
+			return (0);
```
There's no need to do two string comparisons or store the same string with a '/'. Do something similar to this instead (using my recommended rename of the struct field):

```c
const char *fsname = zfs_get_name(zhp);
const size_t parent_len = strlen(state->ga_parent);
boolean_t descendant = true;

if (strncmp(state->ga_parent, fsname, parent_len) == 0) {
	descendant = (fsname[parent_len] == '/' || fsname[parent_len] == '\0');
}

if (!descendant) {
	zfs_close(zhp);
	return (0);
}
```

QORTEC replied:
I did this to resolve the following issue:
````
zpool/dataset/name
zpool/dataset/name2
````

The trailing slash is used to ensure we don't accidentally mount `zpool/dataset/name2` wen the target is `zpool/dataset/name`and its children.

ahesford replied:
You should read my test again.

QORTEC replied:
Don't know why I didn't see the code block when I made the response.

`fsname[parent_len] ` is definitely a better way to accomplish this, thanks for the suggestion.

ahesford commented on cmd/zfs/zfs_main.c:
```diff
@@ -6731,7 +6732,7 @@ zfs_do_holds(int argc, char **argv)
 	boolean_t	ga_verbose;
 	get_all_cb_t	*ga_cbp;
-	const char	*ga_dataset;
+	const char	*ga_filesystem;
```
This should really be called `ga_parent` or something similar, to make it clear that the state is a desired *parent* rather than a desired *filesystem*. I only recommend using the `filesystem` variable name in the `share_mount` function, where you are using it to refer to the "filesystem" argument.

QORTEC replied:
Good point, `parent` is more descriptive/accurate.

ahesford commented on cmd/zfs/zfs_main.c:
```diff
@@ -6787,12 +6807,12 @@ get_one_dataset(zfs_handle_t *zhp, void *data)
 	    .ga_verbose = verbose,
 	    .ga_cbp = cbp,
-	    .ga_dataset = dataset
+	    .ga_filesystem = filesystem
```
As in the struct, `filesystem` here really ought to be `parent` to make it clear you're looking for datasets with a give *parent*.

ahesford commented on cmd/zfs/zfs_main.c:
```diff
@@ -7212,23 +7233,34 @@ share_mount(int op, int argc, char **argv)
+		if (op == OP_MOUNT)
+			filesystem = argv[0];
+		if (filesystem != NULL &&
+		    zfs_open(g_zfs, filesystem, ZFS_TYPE_FILESYSTEM) == NULL) {
```
Doesn't this need to be closed?

ahesford commented on cmd/zfs/zfs_main.c:
```diff
@@ -7212,23 +7233,34 @@ share_mount(int op, int argc, char **argv)
+		 * Validate filesystem is actually a valid zfs filesystem
+		 */
+		const char *filesystem = NULL;
+		if (op == OP_MOUNT)
+			filesystem = argv[0];
```
Is there a reason this is limited only to `mount`?

QORTEC replied:
Yes/no, this pull request was originally for mount only, and `zfs share` has additional args `[smb|nfs]` that I didn't see where it was pulled from.
I thought it best to limit the PR to `zfs mount` untill/unless I look into how share works.

ahesford replied:
There are no additional arguments documented for `zfs share`. The control flow is exactly the same except for the part that actually acts on each filesystem, so limiting this to mount-only seems arbitrary and causes an unnecessary split in the nature of the two recursive filesystem operations.

QORTEC replied:
Agreed, I didn't see anything in the manpages, but usage differs?
Hence why I disabled it until I can investigate it properly.

ahesford replied:
By the time you're consuming this argument, any `nfs` or `smb` argument has already been consumed zfs `zfs share`, so the syntax for `zfs share` becomes
```sh
zfs share [-l] <filesystem | <-a|-r> [nfs|smb] [filesystem]>
```
assuming, of course, that you don't drop the `-r`.

QORTEC replied:
`-r` is not currently enabled for `zfs share`, I was using `op == OP_MOUNT` to restrict `-a [filesystem]`.

I believe this would be the correct syntax for the `-a` _arg_;
````
zfs share [-l] <-a [nfs|smb] [filesystem] | filesystem>
````

My original reasoning for disabling `-a [filesystem]`
- I didn't have time to track down how the args work to verify if my usage is correct
- I would need to update the man pages adding the missing information [nfs|smb]
- I haven't use `zfs share` (something I should do when I have spare time)

ahesford reviewed (commented):
I'll offer some suggestions subject to the caveats that they should be reconciled with whatever stylistic expectations the project expects, and also that you should seek counsel about whether there are more appropriate ways to test the parentage of a filesystem than string comparisons. I would think some `libzfs` routine exists to do this check in a more robust way.

Also, are you actually testing these changes locally?

QORTEC commented on cmd/zfs/zfs_main.c:
```diff
@@ -6772,12 +6773,31 @@ get_one_dataset(zfs_handle_t *zhp, void *data)
+			return (0);
+		}
+
+		/* Add a trailing / to ga_filesystem */
+		char filesystem[strlen(state->ga_filesystem) + 2];
+		strcpy(filesystem, state->ga_filesystem);
+		strcat(filesystem, "/");
```
This was the simplest/cleanest implementation I could come up with so far, I only had time for a few cursory searches on this topic.
Regarding necessity see comment on accidental mounting below, unless I missed something I need that trailing **/**.

QORTEC commented:
> Also, are you actually testing these changes locally?

Yes, though I still need to create a test script with the intention of breaking my implementation to ensure all works as expected.

ahesford commented on cmd/zfs/zfs_main.c:
```diff
@@ -7195,14 +7233,34 @@ share_mount(int op, int argc, char **argv)
 
-		if (argc != 0) {
-			(void) fprintf(stderr, gettext("too many arguments\n"));
+		if (!recursive && argc > 1) {
+			(void) fprintf(stderr, gettext("usage: "
+			    "zfs mount -a [filesystem]\n"));
 			usage(B_FALSE);
 		}
 
+		if (recursive && argc != 1) {
+			(void) fprintf(stderr, gettext("usage: "
+			    "zfs mount -r <filesystem>\n"));
+			usage(B_FALSE);
```
This change should be dropped; it makes the usage text redundant and isn't necessary. Keep the "too many arguments" and let the main usage routine print the actual usage.

QORTEC commented:
Just got some time to work on my PR again. I've cleaned up the implementation based on the feedback received, and noted a couple of changes as well the reasoning behind them below:

---

> -r is a sort of weird `-a` that behaves exactly as `-a` except that it ignores `canmount=noauto` and mounts it anyway.

> I'd submit that the whole notion of `-r` is ill-formed, and you should abandon the new argument and stick to just adding the option for limiting the scope of `-a` to some subtree.

You make a good point that there are certain expectations regarding what recursive behavior should entail, and my implementation may not align with them. I made a poor choice in using `-r` and `recursive` as my flag and descriptors.

Originally, the intention was for `-r filesystem` to act exactly the same as `-a filesystem`, except it would also mount `canmount=noauto` filesystems. To better reflect this operation, I've chosen to use the `-A` flag to signify that it is a modified implementation of `-a`.

Additionally, I've decided to use `mount_noauto` instead of `recursive` as it more accurately describes the code implementation/intention.

The manpage for `zfs mount` also needed to be updated to align with the code implementation:
````
- Mounts the specified ZFS file system and its children.
+ Mounts the specified filesystem and its children, provided they are available.
````
The original wording, which was used for both `-a` and `-r`, falsely implied explicit recursive mounting. The updated wording should allow for a usage where you can execute `zfs mount -a zpool/dataset`, and if `zpool/dataset` has `canmount=off` the result will be the mounting of the filesystem's children instead of returning an explicit error.

---

> The right place for that check is not here, because it's changing the notional behavior of get_one_filesystem

> you'll need to separate the logic between `explicit` and `recursive` so you can properly ignore `canmount=off` for recursive operations as well.

I've updated the `get_one_dataset()` code so that it no longer includes the `canmount=noauto` check. Instead, I've added the `mount_noauto` property to `share_mount_one()`. This change allows us to modify the `ZFS_CANMOUNT_NOAUTO` check by adding `!mount_noauto`, which enables us to treat `canmount=noauto` as if it were `canmount=yes`.

---
Regarding **`zfs share`**; currently, the pull request does not enable `<-a [nfs|smb] [filesystem] | [-A] [nfs|smb] filesystem>`. My original reasoning:
- need to track down the args code to verify if my usage is correct
- need to update the manpages with information regarding `[nfs|smb]` (since it's currently missing).
- need to allocate spare time to verify that the code functions as expected
````
/* Limit `-a filesystem` to mount only */
if (op == OP_MOUNT)
    filesystem = argv[0];
````

---

**Note to self**: 
- need to update the following tests: (I'm failing these tests)
- [ ] `tests/zfs-tests/tests/functional/cli_root/zfs_mount/zfs_mount_009_neg.ksh`
- [ ] `tests/zfs-tests/tests/functional/cli_root/zfs_mount/zfs_mount_011_neg.ksh`
- need to add additional ones to account for the following functionality:
- [ ] `<-a> [filesystem]`
- [ ] `[-A] <filesystem>`
- [ ] `zpool/data` != `zpool/dataset`

QORTEC commented:
#### Failing Tests
````
# DESCRIPTION:
#	Try each 'zfs mount' with inapplicable scenarios to make sure
#	it returns an error. include:
#		* '-a', but also with a specific filesystem.

set -A args "$mountall $TESTPOOL/$TESTFS"

typeset -i i=0
while (( i < ${#args[*]} )); do
	log_mustnot zfs ${args[i]}
	((i = i + 1))
done
````
[`zfs_mount_009_neg.ksh`](../tree/master/tests/zfs-tests/tests/functional/cli_root/zfs_mount/zfs_mount_009_neg.ksh)

````
# DESCRIPTION:
# Verify that zfs mount should fail with bad parameters

set -A badargs "A" "-A" "-" "-x" "-?" "=" "-o *" "-a"

for arg in "${badargs[@]}"; do
	log_mustnot eval "zfs mount $arg $fs >/dev/null 2>&1"
done
````
[`zfs_mount_011_neg.ksh`](../tree/master/tests/zfs-tests/tests/functional/cli_root/zfs_mount/zfs_mount_011_neg.ksh)

It looks like the tests are failing because:
- both tests seem to be executing `zfs mount -a zpool/dataset` and are expecting an error return
- `zfs_mount_011_neg` is executing `zfs mount -A zpool/dataset`, which it is also expecting to fail

The following tests will probably need to be added:
- `-a zroot/dataset`
  - [ ] ensure that `zroot/dataset` and its children are mounted (if allowed)
  - [ ] ensure that children of `zroot/dataset` are mounted (if `zroot/dataset` is not allowed to be mounted)
  - [ ] ensure we don't mount datasets based on partial name match (`zroot/data` != `zroot/dataset`)
- `-A zroot/dataset`
  - [ ] functions the same as `-a` except it treats `canmount=noauto` as `canmount=yes`

**Note to self**:
- tab indentation is 8 spaces not 4

QORTEC commented on tests/zfs-tests/tests/functional/cli_root/zfs_mount/zfs_mount_009_neg.ksh:
```diff
@@ -52,8 +52,8 @@ datasets=""
 	datasets="$datasets $TESTPOOL/$fs"
 done
-
-set -A args "$mountall $TESTPOOL/$TESTFS"
+set -A args "$mountall $datasets" \
```
Missing line brake

ahesford commented:
My initial rebuke of your contribution was based on my view that it is deeply unethical to submit project changes for consideration when you lack understanding of the contribution and are incapable of defending your submission in a thoughtful review process. This has nothing to do, *per se*, with the fact that you consulted "AI" (more aptly and less sensationally called a "correlation engine") to prepare this work. As I explained before, it is your professed lack of understanding that should have prevented you from offering this submission in the first place or, I would have hoped, caused the project owners to reject the submission in reinforcement of the general principles of any intellectual endeavor. My concept of ethics in this arena is deeply influenced by my engagement with open-source software projects, a hobby which can steal time from other personal and professional activities but offers in exchange opportunities for meaningful interactions and personal growth.

Despite my objection, I devoted time to understand your contribution and offer some guidance. This provided me an opportunity to learn a little about the internal workings of ZFS, and I hoped it would be a learning opportunity for you as well.

Your last response was very obviously produced by the same software responsible for the initial submission. After a few introductory sentences written in an informal, human voice consistent with other messages in this thread, the body of the message shifts to a structured, formal style that is characteristic of every interaction I've ever had with a large language model. Rather than developing a deep understanding of your contribution and defending it in your own terms, you seem to continue merely as a conduit to ChatGPT.

I am unwilling to devote further time to interacting with a correlation engine by proxy.

QORTEC commented:
> I am unwilling to devote further time

Thank you for all the time and help you devoted to this pull request.

> human voice consistent with other messages in this thread

If I understand correctly, the messages you are referring to where written using the GitHub Mobile app. Those replies where written while I was on my way to and from work, also I generally don't like to using AI for responses/conversations with others.

> formal style that is characteristic of every interaction I've ever had with a large language model.

I've had a couple of days off, and was able to dedicate more time to working on the PR.
The reason for behind the tone shift is multi-fold;
- I am engaging using a different medium
- my posts where of a different nature 
  - for status updates and thoughts on technical issues, I prefer using formal language

My recent posts/messages have been written with **AI Assistance** as defined by [RoyalRoad AI Text Policy](https://www.royalroad.com/blog/57/royal-road-ai-text-policy).
- I ask the AI to "review my text for spelling and grammar errors"
  - unless I specifically state that "I am happy with the wording" 
    the AI will still continue to suggest alternative wording along with the requested fixes
- I also use AI to help format/express ideas that I had difficulty writing down
  - by providing sentence fragments that are poorly connected the AI can often find a path forwards in express these thoughts in a coherent manner

QORTEC commented on tests/zfs-tests/tests/functional/cli_root/zfs_unmount/zfs_unmount_all_002_pos.ksh:
```diff
@@ -0,0 +1,203 @@
+}
+
+
+log_assert "Verify that 'zfs $unmountall filesystem' succeeds as root, " \
+	"and all available related ZFS filesystems are unmounted."
```
This sentence feels clumsy.

QORTEC commented on tests/zfs-tests/tests/functional/cli_root/zfs_unmount/zfs_unmount_all_002_pos.ksh:
```diff
@@ -0,0 +1,203 @@
+	unset __ZFS_POOL_RESTRICT
+
+	log_must verify_related
+	log_note "Verify that 'zfs $mountcmd' will display " \
+	"all ZFS filesystems related to '$TESTPOOL/$mnt' are currently unmounted."
```
`all ZFS filesystems related to` should be `all available ZFS filesystems related to`

QORTEC commented on tests/zfs-tests/tests/functional/cli_root/zfs_mount/zfs_mount_all_002_pos.ksh:
```diff
@@ -0,0 +1,209 @@
+}
+
+
+log_assert "Verify that 'zfs $mountall filesystem' succeeds as root, " \
+	"and all available related ZFS filesystems are mounted."
```
This sentence feels clumsy.

QORTEC commented on tests/zfs-tests/tests/functional/cli_root/zfs_mount/zfs_mount_all_002_pos.ksh:
```diff
@@ -0,0 +1,209 @@
+
+verify_related true
+
+log_note "Verify that 'zfs $mountcmd' will display " \
+	"all ZFS filesystems related to '$TESTPOOL/$mnt' are currently mounted."
```
`all ZFS filesystems related to` should be `all available ZFS filesystems related to`

QORTEC commented on tests/zfs-tests/tests/functional/cli_root/zfs_unmount/zfs_unmount_all_002_pos.ksh:
```diff
@@ -0,0 +1,203 @@
+
+done
+
+log_pass "'zfs mount -[f]a' succeeds as root, " \
+	"and all available ZFS filesystems are unmounted."
```
needs updating

QORTEC commented on tests/zfs-tests/tests/functional/cli_root/zfs_mount/zfs_mount_all_002_pos.ksh:
```diff
@@ -0,0 +1,209 @@
+
+verify_mount_display
+
+log_pass "'zfs $mountall' succeeds as root, " \
+	"and all available ZFS filesystems are mounted."
```
needs updating

QORTEC commented on tests/zfs-tests/tests/functional/cli_root/zfs_mount/zfs_mount_all_002_pos.ksh:
```diff
@@ -0,0 +1,209 @@
+# - true:	it verifies that the $mnt file system is mounted.
+# - false:	it verifies that the $mnt file system is unmounted.
+#
+# In both scenarios, it ensures that the file systems in fs1 remain unmounted.
```
`fs1` should be `fs`

QORTEC commented on cmd/zfs/zfs_main.c:
```diff
@@ -7574,15 +7645,26 @@ unshare_unmount(int op, int argc, char **argv)
+			/*
+			 * Skip any dataset that's not related to filesystem.
+			 */
+			if (filesystem != NULL &&
+			    !zfs_related_dataset(zhp, filesystem)) {
+				continue;
+			}
```
Brackets are optional `{}`
```c
if (filesystem != NULL &&
    !zfs_related_dataset(zhp, filesystem))
	continue;
```

QORTEC commented on cmd/zfs/zfs_main.c:
```diff
@@ -306,8 +306,8 @@ get_usage(zfs_help_t idx)
 		    "[-s property]...\n\t    [-S property]... [-t type[,...]] "
 		    "[filesystem|volume|snapshot] ...\n"));
 	case HELP_MOUNT:
-		return (gettext("\tmount\n"
-		    "\tmount [-flvO] [-o opts] <-a | filesystem>\n"));
```
Is the double mount purposeful? 
````
usage:
        mount
        mount [-flvO] [-o opts] <-a | filesystem>
````

QORTEC reviewed (commented):
The main issues I've found are comments and log messages that need fixing/updating.
Next I will recommit/rewrite my Git history so the code additions/changes are easier to follow.
Functionally speaking the PR looks good, so I could probably remove the _Draft_ status when these changes are committed.

amotin commented:
This is already fixed by #16015.
